crates/wasmtime/src/runtime/types/matching.rs

```diff
@@ -8,22 +9,35 @@ use wasmtime_environ::{
 use wasmtime_runtime::VMSharedTypeIndex;
 
 pub struct MatchCx<'a> {
-    pub signatures: &'a TypeCollection,
-    pub types: &'a ModuleTypes,
-    pub engine: &'a Engine,
+    signatures: &'a TypeCollection,
+    types: &'a ModuleTypes,
+    engine: &'a Engine,
 }
 
 impl MatchCx<'_> {
-    pub fn vmshared_signature_index(
+    /// Construct a new matching context for the given module.
+    pub fn new(module: &Module) -> MatchCx<'_> {
+        MatchCx {
+            signatures: module.signatures(),
+            types: module.types(),
+            engine: module.engine(),
+        }
+    }
+
+    fn type_reference(
         &self,
         expected: ModuleInternedTypeIndex,
         actual: VMSharedTypeIndex,
     ) -> Result<()> {
         let matches = match self.signatures.shared_type(expected) {
-            Some(idx) => actual == idx,
             // If our expected signature isn't registered, then there's no way
             // that `actual` can match it.
             None => false,
+            Some(expected) => {
+                let expected = FuncType::from_shared_type_index(self.engine, expected);
+                let actual = FuncType::from_shared_type_index(self.engine, actual);
+                actual.matches(&expected)
```

Comment:
Would it be worth adding a special case here to test if `expected == actual` and short-circuit everything else?

Reply:
I'll move that into the `matches` method so every caller bebnefits from it.